File: AlgebraMethod.py
import numpy as np
import unittest


class ElectromagneticField():
    def __init__(self,
                 x,
                 y,
                 z,
                 Mz=1,
                 f=20,
                 sigma=0.1,
                 mu=4 * np.pi * 10**(-7)):
        # !M = 1 A m**2
        # !f = 20 Khz
        self.x = x
        self.y = y
        self.z = z
        self.r = np.sqrt(self.x**2 + self.y**2 + self.z**2)
        self.Mz = Mz
        self.omega = 2 * np.pi * f
        self.mu = mu
        self.sigma = sigma
        self.k = np.sqrt((0 + 1j) * self.omega * self.mu * self.sigma)

    def GetEz(self):
        self.Ez = 0

    def GetEx(self):
        self.Ex = (self.Mz / (4 * np.pi)) * (0 + 1j) * self.omega * self.mu * (
            self.y * np.exp((0 + 1j) * self.k * self.r) / self.r**3) * (
                (0 + 1j) * self.k * self.r - 1)
        print(f"At ({self.x},{self.y},{self.z}) Ex is {self.Ex} ")
        return self.Ex

    def GetEy(self):
        self.Ey = -(self.Mz / (4 * np.pi)) * (
            0 + 1j) * self.omega * self.mu * (self.x * np.exp(
                (0 + 1j) * self.k * self.r) / self.r**3) * (
                    (0 + 1j) * self.k * self.r - 1)
        print(f"At ({self.x},{self.y},{self.z}) Ex is {self.Ex} ")
        return self.Ex


class TestElectromagneticField(unittest.TestCase):
    def setUp(self):
        self.EMF = ElectromagneticField(0.5, 1, 1)

    def testEx(self):
        Ex = self.EMF.GetEx()
        self.assertTrue(np.abs(Ex - 1.0) < 10)


if __name__ == '__main__':
    unittest.main()

File: AlgebraMethod.f90
module AlgebraMethod
    implicit none
    save
    REAL,PARAMETER:: pi=3.14159265359
    REAL,PARAMETER:: Mz =1 !1A m**2
    REAL,PARAMETER:: omega =2 * pi * 20000000
    COMPLEX,PARAMETER :: i=(0,1)
    real,PARAMETER::epsilon=8.854*10**(-12.0)
    complex,PARAMETER:: sigma=0.1-i * omega *epsilon
    REAL,PARAMETER:: mu = 4 * pi * 10**(-7.0)
    COMPLEX :: k=sqrt(i*omega*mu*sigma)
    real::x,y,z
    !!!!!!!!!!
    contains
    subroutine GetPosition(arg1, arg2 , arg3)
        implicit none
        real,intent(in) :: arg1
        real,intent(in) ::  arg2
        real,intent(in) ::  arg3
        x=arg1
        y=arg2
        z=arg3
    end subroutine GetPosition
    subroutine GetEx(Ex)
    implicit none
    REAL :: r 
    COMPLEX,INTENT(OUT)::Ex
    r=sqrt( x**2+y**2+z**2)
    Ex = ( Mz / (4 * pi)) * i *  omega *  mu * (y * exp(i *  k *  r) /  r**3) * (i *  k *  r - 1)
    end subroutine GetEx
    !!!!!!
    subroutine GetEy(Ey)
        implicit none
        REAL :: r 
        COMPLEX,INTENT(OUT)::Ey
        r=sqrt( x**2+y**2+z**2)
        Ey = -( Mz / (4 * pi)) * i *  omega *  mu * ( x * exp(i *  k *  r) /  r**3) * (i *  k *  r - 1)
    end subroutine GetEy
    !!!!!
    subroutine GetEz(Ez)
        COMPLEX,INTENT(OUT)::Ez
        Ez=0*x*y*z
    end subroutine GetEz
    !!!
    subroutine GetH(Hx,Hy,Hz)
        REAL :: r 
        COMPLEX,INTENT(OUT)::Hx,Hy,Hz
        r=sqrt( x**2+y**2+z**2)
        Hx=-((Mz*x*z*exp(i*k*r))/(4*pi))*(k**2/r**3+3*i*k/r**4-3/r**5)
        Hy=-((Mz*y*z*exp(i*k*r))/(4*pi))*(k**2/r**3+3*i*k/r**4-3/r**5)
        Hz=((Mz*exp(i*k*r))/(4*pi*r))*(k**2+i*k/r-(k**2*z**2+1)/r**2-(3*i*k*z**2)/r**3+3*z**2/r**4)
    end subroutine GetH
end module AlgebraMethod




